channels/rdpdr/server/rdpdr_main.c

```diff
@@ -591,37 +825,968 @@ static int rdpdr_server_stop(RdpdrServerContext* context)
+
+	/* Invoke the delete directory completion routine. */
+	context->OnDriveDeleteDirectoryComplete(context, irp->CallbackData, ioStatus);
+
+	/* Destroy the IRP. */
+	rdpdr_server_irp_free(irp);
+}
+
+static void rdpdr_server_drive_delete_directory_callback1(RdpdrServerContext* context, wStream* s, RDPDR_IRP* irp, UINT32 deviceId, UINT32 completionId, UINT32 ioStatus)
+{
+	UINT32 fileId;
+	UINT8 information;
+
+	WLog_DBG(TAG, "%s: deviceId=%d, completionId=%d, ioStatus=0x%x", __FUNCTION__, deviceId, completionId, ioStatus);
+
+	if (ioStatus != STATUS_SUCCESS)
+	{
+		/* Invoke the delete directory completion routine. */
+		context->OnDriveDeleteFileComplete(context, irp->CallbackData, ioStatus);
+
+		/* Destroy the IRP. */
+		rdpdr_server_irp_free(irp);
+
+		return;
+	}
+
+	Stream_Read_UINT32(s, fileId); /* FileId (4 bytes) */
+	Stream_Read_UINT8(s, information); /* Information (1 byte) */
+
+	/* Setup the IRP. */
+	irp->CompletionId = context->priv->NextCompletionId++;
+	irp->Callback = rdpdr_server_drive_delete_directory_callback2;
+	irp->DeviceId = deviceId;
+	irp->FileId = fileId;
+
+	rdpdr_server_enqueue_irp(context, irp);
+
+	/* Send a request to close the file */
+	rdpdr_server_send_device_close_request(context, deviceId, fileId, irp->CompletionId);
+}
+
+static BOOL rdpdr_server_drive_delete_directory(RdpdrServerContext* context, void* callbackData, UINT32 deviceId, const char* path)
+{
+	RDPDR_IRP* irp;
+
+	/* Create an IRP. */
+	irp = rdpdr_server_irp_new();
+	if (irp == NULL) return FALSE;
+
+	irp->CompletionId = context->priv->NextCompletionId++;
+	irp->Callback = rdpdr_server_drive_delete_directory_callback1;
+	irp->CallbackData = callbackData;
+	irp->DeviceId = deviceId;
+	strncpy(irp->PathName, path, sizeof(irp->PathName));
+
+	rdpdr_server_convert_slashes(irp->PathName, sizeof(irp->PathName));
+
+	printf("PathName=%s\n", irp->PathName); fflush(stdout);
+
+	rdpdr_server_enqueue_irp(context, irp);
+
+	/* Send a request to open the file. */
+	rdpdr_server_send_device_create_request(
+		context, deviceId, irp->CompletionId, irp->PathName,
+		DELETE | SYNCHRONIZE,
+		FILE_DIRECTORY_FILE | FILE_DELETE_ON_CLOSE | FILE_SYNCHRONOUS_IO_NONALERT,
+		FILE_OPEN);
+
+	return TRUE;
+}
+
+/*************************************************
+ * Drive Query Directory
+ ************************************************/
+
+static void rdpdr_server_drive_query_directory_callback2(RdpdrServerContext* context, wStream* s, RDPDR_IRP* irp, UINT32 deviceId, UINT32 completionId, UINT32 ioStatus)
+{
+	FILE_DIRECTORY_INFORMATION fdi;
+	UINT32 length;
+
+	WLog_DBG(TAG, "%s: deviceId=%d, completionId=%d, ioStatus=0x%x", __FUNCTION__, deviceId, completionId, ioStatus);
+
+	Stream_Read_UINT32(s, length); /* Length (4 bytes) */
+
+	if (length > 0)
+	{
+		rdpdr_server_read_file_directory_information(s, &fdi);
+	}
+	else
+	{
+		Stream_Seek(s, 1); /* Padding (1 byte) */
+	}
+
```

Comment:
Formatting. By convention in FreeRDP aren't any braces used for one line statements. (if and else in this case)